scripts/setup_ubuntu.sh

```diff
@@ -22,9 +22,9 @@ done
 
 run_as_original_user() {
   if [[ -n "${ORIGINAL_USER}" ]]; then
-    su - "${ORIGINAL_USER}" bash -c "$*"
+    su - "${ORIGINAL_USER}" bash -c '"$@"' -- "$@"
   else
-    bash -c "$*"
+    bash -c '"$@"' -- "$@"
```

Comment:
I need a shell primer: what does repeating `$@` do?

Reply:
The script running in the subshell is `"$@"` which expands to the arguments the script receives without splitting or interpreting the arguments any further. Bash will try to run the whole set of arguments as a command line with the first argument used as the command. The arguments to this script are `"$@"` but in the function scope. These expand to all the arguments without splitting (as above) and pass those as arguments to the script. This means that no splitting will happen after the arguments are split for the call to the function.

I made this fix because I needed to pass `"pip<=20.3"` (or whatever) as an argument to a function called via `run_as_original_user` and the nested shell was splitting the the argument and interpreting the `<` as an input redirect.

Reply:
Basically I was implementing what in C++ they might call "perfect forwarding". Passing through the arguments to a subshell without any change.

This isn't an idiom I found somewhere, by the way. There may be a better way to do it. I just thought this up in the moment.

Reply:
Ah. I had to play around with an example to understand.

The second "$@" is expanded in the script and becomes arguments to a simple "trampoline" script (`bash -c '"$@"'`) that simply executes its command line arguments.

Do you think a comment like:
```bash
# The second "$@" is evaluated in this script and becomes the argument to a bash script `bash -c '"$@"'`
# that simply executes its command-line arguments. This protects characters like < from being interpreted
# as pipe redirections if $@ were stringified directly.
```

would be useful?

Reply:
Sure. I added a version of this that is modified to use the actual shell terminology (to help people search for the ideas). Take a look.